core/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/status/TransportSnapshotsStatusAction.java

```diff
@@ -190,41 +196,49 @@ private SnapshotsStatusResponse buildResponse(SnapshotsStatusRequest request, Li
                     SnapshotIndexShardStatus shardStatus = new SnapshotIndexShardStatus(shardEntry.key, stage);
                     shardStatusBuilder.add(shardStatus);
                 }
-                builder.add(new SnapshotStatus(entry.snapshotId(), entry.state(), Collections.unmodifiableList(shardStatusBuilder)));
+                builder.add(new SnapshotStatus(entry.snapshot(), entry.state(), Collections.unmodifiableList(shardStatusBuilder)));
             }
         }
         // Now add snapshots on disk that are not currently running
-        if (Strings.hasText(request.repository())) {
-            if (request.snapshots() != null && request.snapshots().length > 0) {
-                for (String snapshotName : request.snapshots()) {
-                    SnapshotId snapshotId = new SnapshotId(request.repository(), snapshotName);
-                    if (currentSnapshotIds.contains(snapshotId)) {
-                        // This is a snapshot the is currently running - skipping
-                        continue;
+        final String repositoryName = request.repository();
+        if (Strings.hasText(repositoryName) && request.snapshots() != null && request.snapshots().length > 0) {
+            final Set<String> requestedSnapshotNames = Sets.newHashSet(request.snapshots());
+            final Set<String> currentSnapshotNames = currentSnapshotIds.stream().map(SnapshotId::getName).collect(Collectors.toSet());
+            final Map<String, SnapshotId> matchedSnapshotIds = snapshotsService.snapshotIds(repositoryName).stream()
+                .filter(s -> requestedSnapshotNames.contains(s.getName()) && currentSnapshotIds.contains(s) == false)
```

Comment:
I wonder if we can simplify this. What we currently achieve here is that we prioritize running snapshots over stored ones (with same name). If that's what we want, we can instead of creating `currentSnapshotIds` at the beginning of the method create a `Set<String> currentSnapshotNames`. Instead of `matchedSnapshotIds` we just use `Map<String, SnapshotId> storedSnapshotIds` which are just all  the stored snapshots. Using the check `if (currentSnapshotNames.contains(snapshotName)) {` below we then get the same behavior as now?